### ISSUE TITLE
Concatenating CSR matrices fails

### ISSUE THREAD

**Issue by rootjalex** [pytorch/pytorch#129843]
### 🐛 Describe the bug

The [torch.sparse documentation](https://pytorch.org/docs/stable/sparse.html#other-functions) claims to support (among others), `cat`, `stack`, `vstack`, and `hstack`. None of these functions work on CSR matrices, failing with `RuntimeError: Sparse CSR tensors do not have is_contiguous`.

```python
import torch

crow_indices = torch.tensor([0, 2, 4], dtype=torch.int32)
col_indices = torch.tensor([0, 1, 0, 1], dtype=torch.int32)
values = torch.tensor([1, 2, 3, 4], dtype=torch.float32)
A = torch.sparse_csr_tensor(crow_indices, col_indices, values, dtype=torch.float32)

# Each of these fail
print(torch.hstack((A, A)))
print(torch.vstack((A, A)))
print(torch.cat((A, A), dim=0))
print(torch.cat((A, A), dim=1))
print(torch.stack((A, A)))
```
Here is the backtrace for the first failure:
```bash
% python3 file.py
file.py:6: UserWarning: Sparse CSR tensor support is in beta state. If you miss a functionality in the sparse tensor support, please submit a feature request to https://github.com/pytorch/pytorch/issues. (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/SparseCsrTensorImpl.cpp:55.)
  A = torch.sparse_csr_tensor(crow_indices, col_indices, values, dtype=torch.float32)
Traceback (most recent call last):
  File "file.py", line 9, in <module>
    print(torch.hstack((A, A)))
          ^^^^^^^^^^^^^^^^^^^^
RuntimeError: Sparse CSR tensors do not have is_contiguous
```

### Versions

Collecting environment information...
PyTorch version: 2.3.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.4.1 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.3.9.4)
CMake version: version 3.29.2
Libc version: N/A

Python version: 3.12.3 (main, Apr  9 2024, 08:09:14) [Clang 15.0.0 (clang-1500.1.0.2.5)] (64-bit runtime)
Python platform: macOS-14.4.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] numpy==2.0.0
[pip3] torch==2.3.1
[conda] Could not collect

**Comment by alireza78h** [comment]
Hey, any updates on this?

**Comment by jcaip** [comment]
I'll take a deeper look at this sometime this week, but I don't think these `cat` operations are implemented and the docs are out of date. Off the top of my head though, seems like you should be able to do this by messing with the crow, col, and values tensors. 

@zbh2047 Curious if you can share more on your use case for CSR tensors? I work on sparsity for torchao and am generally interested in how people are using it.

**Comment by zbh2047** [comment]
The use case for me is that the 2d matrix represents a “dataset” where each row is a sample, which is sparse and has lots of zeros. Concatenating these matrices along the sample dimension is quite natural in this case.

Another use case is that if each matrix corresponds to the “adjacency” matrix in a bipartite graph, then concatenating matrices means merging two bipartite graphs that share the column vertices but have different row vertices.

For both use cases, CSR representation is quite natural so I think implementing cat for CSR is necessary. Moreover, I feel that implementing concatenate directly on CSR is not hard? Just need a cumsum for crow, and simply concatenate col and values. Please correct me if I am wrong. @jcaip @pearu

**Comment by pearu** [comment]
> Moreover, I feel that implementing concatenate directly on CSR is not hard? Just need a cumsum for crow, and simply concatenate col and values.

The complexity of concatenate implementation depends on how the input data aligns with the desired output.

For example, concatenation of CSR tensors along compressed (row) dimension is as simple as you described, column indices and values are simply stacked and compressed indices are stacked with an offset:
```python
>>> x
tensor(crow_indices=tensor([0, 1, 3]),
       col_indices=tensor([0, 0, 1]),
       values=tensor([1, 2, 3]), size=(2, 2), nnz=3, layout=torch.sparse_csr)
>>> y
tensor(crow_indices=tensor([0, 1, 2]),
       col_indices=tensor([1, 0]),
       values=tensor([4, 5]), size=(2, 2), nnz=2, layout=torch.sparse_csr)
>>> torch.cat((x.to_sparse_coo(), y.to_sparse_coo()), dim=0).to_sparse_csr()
tensor(crow_indices=tensor([0, 1, 3, 4, 5]),
       col_indices=tensor([0, 0, 1, 1, 0]),
       values=tensor([1, 2, 3, 4, 5]), size=(4, 2), nnz=5,
       layout=torch.sparse_csr)
```
or when using CSR data directly:
```python
>>> new_compressed_indices = torch.cat((x.crow_indices(), x.crow_indices()[-1] + y.crow_indices()[1:]))
>>> new_plain_indices = torch.cat((x.col_indices(), y.col_indices()))
>>> new_values = torch.cat((x.values(), y.values()))
>>> torch.sparse_csr_tensor(new_compressed_indices, new_plain_indices, new_values)
tensor(crow_indices=tensor([0, 1, 3, 4, 5]),
       col_indices=tensor([0, 0, 1, 1, 0]),
       values=tensor([1, 2, 3, 4, 5]), size=(4, 2), nnz=5,
       layout=torch.sparse_csr)
```
I think this case is simple enough and worth implementing in pytorch.

However, concatenation of CSR tensors along plain (column) dimension is not as simple because the new indices and values consists of a mix of inputs indices and values:
```python
>>> torch.cat((x.to_sparse_coo(), y.to_sparse_coo()), dim=1).to_sparse_csr()
tensor(crow_indices=tensor([0, 2, 5]),
       col_indices=tensor([0, 3, 0, 1, 2]),
       values=tensor([1, 4, 2, 3, 5]), size=(2, 4), nnz=5,
       layout=torch.sparse_csr)
```
To support this case, I think using the intermediate COO tensors concatenation is a best approach because the complexity of the merging algorithm for indices/values is going to be the same as when trying to use CSR data directly.

**Comment by zbh2047** [comment]
Yes, concatenating along rows is exactly what I mean. I think this is the most important use case as described previous examples. Concatenating along columns is indeed nontrivial and may be less useful. The use of intermediate coo sparse tensors could already be good for this case.

**Comment by rootjalex** [comment]
Stacking CSR matrices along columns is not extremely complicated, and can be implemented with simple loops like:
```
// C = hstack(A, B)
for (size_t i = 0; i < A_rows; i++) {
  // copy a row of A into C
  for (size_t jp = A_crows[i]; jp < A_crows[i + 1]; jp++) {
    C_crds[cp] = A_crds[jp];
    C_data[cp++] = A_data[jp];
  }
  // copy a row of B into C
  for (size_t jp = B_crows[i]; jp < B_crows[i + 1]; jp++) {
    C_crds[cp] = A_N_cols + B_crds[jp];
    C_data[cp++] = B_data[jp];
  }
  // 
  C_crows[i] = cp;
}
```

Sorry for the shameless self-plug, but we wrote [a paper on how to generate shape operators for various sparse tensor formats](https://ajroot.pl/publications/oopsla2024-burrito.pdf), which could be a useful starting point for implementing a lot of these. [The software artifact](https://github.com/rootjalex/burrito-artifact) generates C++ code that also might be a nice starting point for implementations, though is definitely not perfect (e.g. always generates `while` loops even when it really should generate `for` loops), but has been pretty thoroughly tested on CSR/CSC/COO for various `reshape`/`concat`/`slice` operations.

**Comment by pearu** [comment]
Thanks, @rootjalex , for sharing the paper. From a quick read, this approach is well applicable for CPU kernels but since PyTorch is a CUDA oriented array library, using algorithms that support concurrent workflows are preferred, or algorithms that can be formulated using PyTorch operations that use optimized CPU/CUDA kernels.

**Comment by rootjalex** [comment]
@pearu Yeah, unfortunately, it is entirely CPU-based code generation. I think writing to (general) sparse outputs in parallel is still a fairly open research problem, I'm not familiar with any open source GPU sparse-outputs work yet.

In terms of implementing these operations via already optimized array kernels, I believe scipy.sparse implements this operation via just a series of numpy kernels: https://github.com/scipy/scipy/blob/d1073acbc804b721cfe356969d8461cdd25a7839/scipy/sparse/_construct.py#L656-L709
This is slower than the above (fused) code on a single processor, but might be more amenable to what you're looking for.

**Comment by freshentropy** [comment]
I have some free time next week so I'd like to take a stab at this if nobody's actively on it.

dim=0 is what @pearu described above. For dim=1 I don't think we actually need a per row merge pass. The crow arrays are already prefix sums, so

```
crow_C = crow_A + crow_B
dest of A's k-th element = k + crow_B[row]
dest of B's k-th element = k + crow_A[row+1]   (cols shifted by A.size(1))

```
I think this can be done with existing ops (repeat_interleave/searchsorted + index_copy), with the same path for CPU and CUDA.

I'll keep it to plain 2D CSR and cat on dim 0 and 1, which should also cover vstack/hstack. I'd leave stack/dstack, CSC/BSR, and autograd for later. I'd also fix the docs list that currently says this works.

If anyone sees an issue with this approach, lmk.

@jcaip you're assigned. Is it okay if I pick this up?

**Comment by jcaip** [comment]
Hi @alxhosse I left pytorch this april (and no longer have permissions to unassign myself), so I can't speak to whether a fix would be accepted, I would check with a current maintainer.

**Comment by freshentropy** [comment]
Thanks @jcaip. @pearu @amjames @nikitaved any objection to me picking this up with the approach above? Planning to start next week
